proposals/pattern-const-expressions.md

```diff
@@ -0,0 +1,246 @@
+> 
+> - Constant declarations ([§15.4](classes.md#154-constants))
+> - Enumeration member declarations ([§19.4](enums.md#194-enum-members))
+> - Default arguments of formal parameter lists ([§15.6.2](classes.md#1562-method-parameters))
+> - `case` labels of a `switch` statement ([§13.8.3](statements.md#1383-the-switch-statement)).
+> - `goto case` statements ([§13.10.4](statements.md#13104-the-goto-statement))
+> - Dimension lengths in an array creation expression ([§12.8.16.5](expressions.md#128165-array-creation-expressions)) that includes an initializer.
+> - Attributes ([§22](attributes.md#22-attributes))
+> - In a *constant_pattern* ([§11.2.3](patterns.md#1123-constant-pattern))
+
+Based on the above, the contexts that will accept constant `is` pattern expressions are constant contexts that accept `bool` values, namely:
+- Constant declarations
+- Default arguments of formal parameter lists
+- Attribute arguments
+- `case` labels of a `switch` statement
+- `goto case` statements
+- Switch expression arms
+
+Examples for the above:
+```csharp
+const int a = 4;
+const bool b = false;
+const int c = 4;
+const string d = "hello";
+const Sign e = Sign.Negative;
+
+// constant declaration
+const bool x = a is default(int); // always false, no warning/error
+
+// attribute argument
+[assembly: Something(a is 3)] // always false, no warning/error
+
+// default argument of formal parameter list
+// always true, no warning/error
+int Negate(int value, bool negate = e is Sign.Negative) { }
+
+// switch statement case label
+// + goto case statement
+switch (b)
+{
+    // always false, no warning/error
+    case a is not c:
+        break;
+
+    default:
+        // always false, no warning/error
+        goto case a is not c;
+}
+
+// switch expression arm
+var p = b switch
+{
+    // always true, no warning/error
+    a is c => 1,
+};
+```
+
+**NOTE**: we do not introduce any breaking changes in the reported diagnostics. Currently, all the above cases are illegal reporting "CS0150: A constant value is expected". A warning about the values always evaluating to either true or false is also reported alongside the error. We remove the warnings from those places where `is` expressions are currently not permitted to be used due to the error.
+
+### Other patterns
+[other-patterns]: #other-patterns
+
+Pattern expressions containing non-constant subpatterns, like accessing properties, list patterns and var patterns, are **not** constant. In the below examples, all expressions will report compiler errors:
+
+```csharp
+const bool q = d is string { Length: 5 }; // Error: not a constant expression
+const bool r = d is [.. var prefix, 'l', 'o']; // Error: not a constant expression
+const bool s = d is var someString; // Error: not a constant expression
+```
+
+### Switch expressions
+[switch-expressions]: #switch-expressions
+
+When evaluating a constant value on a switch expression, we adjust the reported diagnostics based on the matching arms:
```

Comment:
This doesn't seem quite right, and LDM was concerned by the specific behavior that you show examples of below. We wanted to limit the subsumption changes to only constant contexts.

Reply:
Is it fine if we make raise the warning into an error if we fail to match any subpattern in the constant switch? The constant value cannot be assigned without being provided a value from the default case, should all others fail to match.

For example,

```csharp
const int a = 4;
// ERROR: No subpattern is matched, and the assigned value is a constant
const int c = a switch
{
    1 => 2,
    2 => 3,
    3 => 5,
};

// WARNING: No subpattern is matched, the default case will be handled and it will throw at runtime.
int v = a switch
{
    1 => 2,
    2 => 3,
    3 => 5,
};
```

I believe that both of the above cases should be an error, but I'd like your input about those.

Reply:
The second case should have unchanged behavior from today. The first case would indeed be an error.

Reply:
In general, the wording itself is not what I expected here. I don't see any mention of constant contexts, which is what I am expecting to drive changes in warning behavior.